From e4d24fd82009b83952ba7b4e664cda8acb18643d Mon Sep 17 00:00:00 2001
From: Olivier Fourdan <ofourdan@redhat.com>
Date: Wed, 2 Nov 2016 11:51:54 +0100
Subject: [PATCH] wayland: check valid pending cairo surface

drop_cairo_surfaces() will nullify the drop_cairo_surfaces() (to avoid
reusing the buffer later), but gdk_window_impl_wayland_end_paint()
doesn't check if the drop_cairo_surfaces is still valid prior to calling
gdk_wayland_window_attach_image().

As gdk_wayland_window_attach_image() assumes there is a valid pending
cairo surface, make sure the staging cairo surface is stil lvalid prior
to attaching the image, just like it's done in gdk_wayland_window_show()
as well.

https://bugzilla.gnome.org/show_bug.cgi?id=773274
---
 gdk/wayland/gdkwindow-wayland.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/gdk/wayland/gdkwindow-wayland.c b/gdk/wayland/gdkwindow-wayland.c
index a8e6785..150a4df 100644
--- a/gdk/wayland/gdkwindow-wayland.c
+++ b/gdk/wayland/gdkwindow-wayland.c
@@ -904,7 +904,9 @@ gdk_window_impl_wayland_end_paint (GdkWindow *window)
   cairo_rectangle_int_t rect;
   int i, n;
 
-  if (!window->current_paint.use_gl &&
+  if (impl->staging_cairo_surface &&
+      _gdk_wayland_is_shm_surface (impl->staging_cairo_surface) &&
+      !window->current_paint.use_gl &&
       !cairo_region_is_empty (window->current_paint.region))
     {
       gdk_wayland_window_attach_image (window);
-- 
2.9.3

